go-controller/pkg/node/node_ip_handler_linux.go

```diff
@@ -163,24 +198,30 @@ func (c *addressManager) isValidNodeIP(addr net.IP) bool {
 func (c *addressManager) sync() {
 	addrs, err := net.InterfaceAddrs()
 	if err != nil {
-		klog.Errorf("Failed to initialize Node IP Manager: unable list all IPs on the node, error: %v", err)
+		klog.Errorf("Failed to sync Node IP Manager: unable list all IPs on the node, error: %v", err)
+		return
 	}
 
-	addrChanged := false
+	currAddresses := sets.NewString()
 	for _, addr := range addrs {
 		ip, _, err := net.ParseCIDR(addr.String())
 		if err != nil {
 			klog.Errorf("Invalid IP address found on host: %s", addr.String())
 			continue
 		}
-		addrChanged = c.addAddr(ip) || addrChanged
+		if !c.isValidNodeIP(ip) {
+			klog.V(5).Info("Skipping invalid IP address found on host: %s", ip.String())
+			continue
+		}
+		currAddresses.Insert(ip.String())
 	}
 
-	if addrChanged {
+	addrChanged := c.assignAddresses(currAddresses)
+	if addrChanged || !c.doesNodeHostAddressesMatch() {
+		klog.Infof("Node address annotation being set to: %v addrChanged: %v", currAddresses, addrChanged)
 		if err := util.SetNodeHostAddresses(c.nodeAnnotator, c.addresses); err != nil {
 			klog.Errorf("Failed to set node annotations: %v", err)
-		}
-		if err := c.nodeAnnotator.Run(); err != nil {
+		} else if err := c.nodeAnnotator.Run(); err != nil {
 			klog.Errorf("Failed to set node annotations: %v", err)
```

Comment:
NIT: L 223 and 225 have same error messages, so hard to tell from the log file very it failed.. (unless we also look for line numbers)

Reply:
ack. good point. It would be better not to rely on line number. I will tweak that should I need to change this file on another pr.